pkg/liquidity-mining/contracts/L2BalancerPseudoMinter.sol

```diff
@@ -0,0 +1,134 @@
+     * Reverts if the given factory was added beforehand; emits `GaugeFactoryAdded` event upon success.
+     */
+    function addGaugeFactory(ILiquidityGaugeFactory factory) external authenticate {
+        require(!_validFactories[factory], "FACTORY_ALREADY_ADDED");
+        _validFactories[factory] = true;
+        emit GaugeFactoryAdded(factory);
+    }
+
+    /**
+     * @notice Removes a given child chain gauge factory from the allowlist.
+     * @dev This is a permissioned function.
+     * Reverts if the given factory had not been added beforehand; emits `GaugeFactoryRemoved` event upon success.
+     */
+    function removeGaugeFactory(ILiquidityGaugeFactory factory) external authenticate {
+        require(_validFactories[factory], "FACTORY_NOT_ADDED");
+        _validFactories[factory] = false;
+        emit GaugeFactoryRemoved(factory);
+    }
+
+    /**
+     * @notice Returns true if the given child chain gauge factory is in the allowlist; false otherwise.
+     */
+    function isValidGaugeFactory(ILiquidityGaugeFactory factory) public view returns (bool) {
+        return _validFactories[factory];
+    }
+
+    // Internal functions
+
+    function _mintFor(address gauge, address user) internal override returns (uint256 tokensToMint) {
+        tokensToMint = _updateGauge(gauge, user);
+        _pseudoMint(user, tokensToMint);
+    }
+
+    function _mintForMany(address[] calldata gauges, address user) internal override returns (uint256 tokensToMint) {
+        uint256 length = gauges.length;
+        for (uint256 i = 0; i < length; ++i) {
+            tokensToMint = tokensToMint.add(_updateGauge(gauges[i], user));
+        }
+        _pseudoMint(user, tokensToMint);
+    }
+
+    /**
+     * @dev Checkpoints given gauge and updates the internal accounting with the total tokens that should be transfered
+     * to the user since the start.
+     *
+     * @param gauge Gauge to checkpoint and query for total tokens to be transferred.
+     * @param user User to query in the given gauge.
+     * @return tokensToMint Amount of tokens to be transferred to the user, calculated as the difference between the
+     * total amount of tokens as indicated by the gauge and the tokens that have already been transferred to the user.
+     */
+    function _updateGauge(address gauge, address user) internal returns (uint256 tokensToMint) {
+        // First, we retrieve the factory address registered from the gauge.
+        // If the factory address is allowlisted in this contract, we verify that the gauge was actually created by
+        // the factory (otherwise it could be just a malicious gauge that claims to be created by an allowed factory).
+        IChildChainGauge ccGauge = IChildChainGauge(gauge);
+        ILiquidityGaugeFactory factory = ccGauge.factory();
+        require(isValidGaugeFactory(factory), "INVALID_GAUGE_FACTORY");
+        require(factory.isGaugeFromFactory(gauge), "INVALID_GAUGE");
```

Comment:
I think handling permissions this way is fine.

An alternative way would be to have an 'addGauge' function here that is permissioned, and then grant permission to the factory to call it. The problem with that is we'd need to assign a delay to granting said permission, which would have parallels with the GaugeController's add_gauge, but there's no gauge controller here (also I find delay on grant is harder to understand).

I'd rather we keep it like this, same as the L1 (someone needs to register new factories).

Reply:
+1; I also find it more intuitive.
The difference with the L1 is the back and forth between the gauge and the factory (i.e. the gauge needs to hold a reference to the factory so that the pseudo minter can then go and query the factory). But I think that is reasonable, and the overall system should be robust.